src/wallet/wallet.cpp

```diff
@@ -3968,6 +4039,25 @@ CWallet* CWallet::CreateWalletFromFile(const std::string walletFile)
 
     RegisterValidationInterface(walletInstance);
 
+    // Make sure we have enough keys in our keypool if HD is enabled
+    if (walletInstance->IsHDEnabled()) {
+        unsigned int keypool_size = GetArg("-keypool", DEFAULT_KEYPOOL_SIZE);
+        unsigned int keypool_critical = GetArg("-keypoolcritical", DEFAULT_KEYPOOL_CRITICAL);
+        if (walletInstance->IsCrypted()) {
+            if (keypool_size < keypool_critical) {
+                LogPrintf("Parameter Interaction: keypool size (%d) must be larger than keypool critical size for encrypted wallets (%d)\n", keypool_size, keypool_critical);
+                ForceSetArg("-keypool", std::to_string(keypool_critical));
+            }
+        } else {
+            if (keypool_size < keypool_critical) {
+                InitWarning(strprintf(_("Your keypool is configured to store %d keys, which is below the keypool critical size of %d. Using a larger keypool will make it less "
```

Comment:
If we hit this warning, then the wallet is not encrypted, and `TopUpKeyPool()` should succeed.

If the wallet is encrypted we'll hit the branch above and we force set keypool to keypool_critical.